### PR TITLE
Streaming support

### PR DESCRIPTION
**Pull request by edsu**
This commit modifies the `/api/completion` endpoint so that it streams results back from the underlying LLM as JSON-Lines (line delimited JSON objects).

The first object contains information about the LLM request so that the page can be updated with this information. Subsequent lines contain objects that contain a partial response from the LLM.

The client then streams the results back and displays them as they become available.

https://github.com/harvard-lil/warc-gpt/assets/33829/ba8d9648-47b6-4c78-8f93-34a085dd3310

### CODE REVIEW

File: warc_gpt/static/index.js [hunk 1]
```diff
@@ -39,7 +39,7 @@ const askButton = document.querySelector("#ask");
  * @returns {string}
  */
 const sanitizeString = (string, convertLineBreaks = true) => {
-  string = string.trim()
+  string = string
```
Comment (edsu) [inline warc_gpt/static/index.js:42]: trim() needed to be removed here to preserve space between words as they come in

**Comment by matteocargnelutti** [comment]
@edsu Thank you _so much_ for initiating this PR, really appreciate it. Streaming is definitely a great "quality of life" feature to have. 

**I have made a few comments but here is an overview of where I think we are:**
- There are a few tweaks needed to make this feature interoperable. 
- The history feature broke in the process and needs fixing. You can test this by asking a follow-up question and check the contents of the `history` object sent to the API.
- (TBD) At higher level, I wonder if it could be worth decoupling text completion from context / metadata retrieval. 
  - At the moment, everything goes through `[POST] /api/completion` maybe we could have two different routes: 
    - One to handle vector search and history. It would be called first.
    - One to handle text completion, and only that. 
  - That could also be the opportunity to decouple that logic at front-end level. We might need to handle history a little bit differently as a result.
  
Cheers,

**Comment by edsu** [comment]
Thanks yes, I thought about a cleaner API approach by decoupling endpoints. I think that would involve some sort of persistence layer separate from chromadb? That seemed like a big change so I was reluctant to go there :-)

I'll take a look and see if I can figure out what happened with the history. It is fun to see the response come back in dribs and drabs, especially when running with a local model, which can take some time.

**Comment by matteocargnelutti** [comment]
@edsu I added an implementation of streaming in #5, based on our conversation here. Thanks for kicking this off!

**Comment by edsu** [comment]
Awesome, I'm glad my noodling around was helpful in some way.

**Pull request closed without merging**